src/@types/index.ts

```diff
@@ -0,0 +1,375 @@
+     * call to {@link Provider.fetchSessionConfig}
+     */
+    setSessionConfig(value: SessionConfig): void
+
+    /**
+     * A hook called very early by the core module, enabling providers an
+     * opportunity to determine if they should participate in the session,
+     * e.g. based on user agent feature detection, random downsampling, etc.
+     * It returns a promise, in case a provider would like to make a network
+     * request in order to make this determination.
+     */
+    shouldRun(): Promise<boolean>
+
+    /**
+     * @remarks
+     * A provider implements this in order to define its logic for producing
+     * a {@link SessionConfig} object at runtime.
+     */
+    fetchSessionConfig(): Promise<SessionConfig>
+
+    /**
+     * A provider implements this in order to define its logic for converting
+     * the configuration from a {@link SessionConfig} object into one or more
+     * {@link Executable} objects (usually {@link Fetch} or other classes
+     * inheriting from {@link Test}).
+     */
+    expandTasks(): Executable[]
+
+    /**
+     * @remarks
+     * A provider implements this in order to define its logic for creating
+     * a {@link ResultBundle} describing the outcome of running a {@link Fetch}
+     * test.
+     * @param timingEntry The Resource Timing entry used to generate the test
+     * result.
+     * @param response The **Response** object resulting from the fetch
+     * activity.
+     * @param testConfig The test configuration.
+     * @param setupResult The provider-defined result of any test setup
+     * activity done prior to the fetch.
+     */
+    createFetchTestResult(
+        timingEntry: ResourceTimingEntry,
+        response: Response,
+        testConfig: TestConfiguration,
+        setupResult: TestSetupResult,
+    ): Promise<TestResultBundle>
+
+    /**
+     * @remarks
+     * A provider implements this in order to create a beacon payload.
+     * @param testConfig The test configuration.
+     * @param testData The data resulting from running the test.
+     */
+    makeBeaconData(
+        testConfig: TestConfiguration,
+        testData: TestResultBundle,
+    ): Beacon.Data
+
+    /**
+     * A hook enabling a provider to determine the beacon URL for a test
+     * result.
+     * @param testConfig The test configuration.
+     */
+    makeBeaconURL(testConfig: TestConfiguration): string
+
+    /**
+     * A hook enabling a provider to generate a test URL at runtime.
+     * @remarks
+     * A provider may cache the result of the first call to this method and
+     * reuse the value on subsequent calls.
+     * @param testConfig The test configuration, which usually specifies a base
+     * URL from which the provider produces the runtime URL.
+     */
+    getResourceUrl(testConfig: TestConfiguration): URL
+
+    /**
+     * A hook enabling a provider to draw a set of zero or more
+     * {@link HttpHeader} tuples from a provider-defined test configuration.
```

Comment:
This was slightly confusing to me. Maybe:

```suggestion
     * A hook enabling a provider to define a list of zero or more request
     * {@link HttpHeader} tuples to be sent with the test, given a 
     * provider-defined test configuration.
```